packages/block-library/src/query-title/edit.js

```diff
@@ -12,43 +12,91 @@ import {
 	BlockControls,
 	useBlockProps,
 	Warning,
+	RichText,
 } from '@wordpress/block-editor';
-import { __ } from '@wordpress/i18n';
+import { __, _x } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import HeadingLevelDropdown from '../heading/heading-level-dropdown';
 
-const SUPPORTED_TYPES = [ 'archive' ];
+const SUPPORTED_TEMPLATES = [ 'archive', 'search', '404' ];
 
 export default function QueryTitleEdit( {
-	attributes: { type, level, textAlign },
+	attributes: { content, level, textAlign },
 	setAttributes,
+	context: { templateSlug },
 } ) {
 	const TagName = `h${ level }`;
 	const blockProps = useBlockProps( {
-		className: classnames( {
+		className: classnames( 'wp-block-query-title__placeholder', {
 			[ `has-text-align-${ textAlign }` ]: textAlign,
-			'wp-block-query-title__placeholder': type === 'archive',
 		} ),
 	} );
-	// The plan is to augment this block with more
-	// block variations like `Search Title`.
-	if ( ! SUPPORTED_TYPES.includes( type ) ) {
+	let titleElement;
+
+	// Infer title content from template slug context
+	// Defaults to content attribute prop
+	switch ( templateSlug ) {
+		case 'archive':
+			// translators: Title for archive template.
+			content = _x( 'Archive title', 'archive template title' );
+			titleElement = <TagName { ...blockProps }>{ content }</TagName>;
+			break;
+		case 'search':
+			// translators: Title for search template with dynamic content placeholders.
+			const placeholderContent = _x(
+				'Search results for "%search%"',
+				'search template title'
+			);
+
+			// Use placeholder content if new Query Title block is added
+			if ( content === '' ) {
```

Comment:
We shouldn't add the placeholder to content here as the user will not be able to clear the input and write something new.

Reply:
Ah yeah I can see this problem when I try to clear the input. If I use a non-empty default placeholder such as 'Query title' rather than checking for an empty string, it looks like I can clear the input safely from the editor. 

I'm not sure this is the correct fix though, as this logic was mainly because my use of `useEffect` was overwriting the custom content when a new query title block was added. I'll try and fix this as part of a `useEffect` refactor.